src/discord/gateway.clj

```diff
@@ -271,28 +274,31 @@
    receive-channel : Channel -- An asynchronous channel (core.async) that messages from the server
       will be pushed onto."
   [auth receive-channel]
-  (let [socket              (atom nil)
-        seq-num             (atom 0)
-        heartbeat-interval  (atom 1000)
-        session-id          (atom nil)
-        gateway             (build-gateway (http/get-bot-gateway auth))
-        gateway             (assoc gateway
-                                   :auth                auth
-                                   :session-id          session-id
-                                   :seq-num             seq-num
-                                   :heartbeat-interval  heartbeat-interval
-                                   :receive-channel     receive-channel
-                                   :websocket           socket)
-        websocket           (create-websocket gateway)]
+  (let [socket                 (atom nil)
+        seq-num                (atom 0)
+        heartbeat-interval     (atom 1000)
+        stop-heartbeat-channel (async/chan)
+        session-id             (atom nil)
+        gateway                (build-gateway (http/get-bot-gateway auth))
+        gateway                (assoc gateway
+                                      :auth                   auth
+                                      :session-id             session-id
+                                      :seq-num                seq-num
+                                      :heartbeat-interval     heartbeat-interval
+                                      :stop-heartbeat-channel stop-heartbeat-channel
+                                      :receive-channel        receive-channel
+                                      :websocket              socket)
+        websocket              (create-websocket gateway)]
 
     ;; Assign the connected websocket to the Gateway's socket field
     (reset! socket websocket)
 
     ;; Begin asynchronously sending heartbeat messages to the gateway
     (go-loop []
       (send-heartbeat gateway seq-num)
-      (<! (async/timeout @heartbeat-interval))
-      (recur))
+      (let [[_ port] (async/alts! [stop-heartbeat-channel (async/timeout @heartbeat-interval)])]
```

Comment:
I'd never thought about using an `async/alts!` call for this, really 👍 solution to the problem!

I was playing around with this locally and this version is personally a little easier to understand (for me). Curious on your thoughts?

```clojure
(async/alt!
  stop-heartbeat-channel (timbre/warn "Websocket closed! Terminating heartbeat channel...")
  (async/timeout @heartbeat-interval) (recur))
```

Reply:
I like your version better! Thanks for the suggestion. I tried something similar first, but was getting some sass about "recur has to be in the tail position." Looks like I had a syntax error somewhere that I didn't try hard enough to track down.